persistence/data_models_json_fields.go

```diff
@@ -91,6 +91,164 @@ func (s *StateExecutionSequenceMapsJson) CompleteNewStateExecution(stateId strin
+
+func (s *StateExecutionWaitingQueuesJson) AddNewLocalQueueCommandForStateExecution(
+	stateExecutionId StateExecutionId, command xdbapi.LocalQueueCommand) {
+	if command.GetCount() == 0 {
+		command.Count = xdbapi.PtrInt32(1)
+	}
+
+	stateExecutionIdKey := stateExecutionId.GetStateExecutionId()
+
+	s.StateToCommandsMap[stateExecutionIdKey] = append(s.StateToCommandsMap[stateExecutionIdKey], command)
+}
+
+// Consume return (StateExecutionId string, dedupIds) where the StateExecutionId consumes messages of these dudupIds.
+//
+// E.g., given StateToCommandsMap as:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+//	state_3, 1: (q1: 1), (q2, 2)
+//
+// If receiving the queue `q1`, then state_3, 1 will consume the queue `q1`, and StateToCommandsMap becomes:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+//	state_3, 1: (q2, 2)
+//
+// If receiving the queue `q2` for twice, then state_3, 1 will consume the two queue `q2`, and StateToCommandsMap becomes:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+func (s *StateExecutionWaitingQueuesJson) Consume(message LocalQueueMessageInfoJson) (*string, []uuid.UUID) {
+	s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] = append(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName], message.DedupId)
+
+	for stateExecutionIdKey, commands := range s.StateToCommandsMap {
+		for i, command := range commands {
+			if command.GetQueueName() != message.QueueName || int(command.GetCount()) > len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) {
+				continue
+			}
+
+			consumedDedupIds := s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][:int(command.GetCount())]
+
+			s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] =
+				s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][int(command.GetCount()):]
+			if len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) == 0 {
+				delete(s.UnconsumedMessageQueueDedupIdsMap, message.QueueName)
+			}
+
+			s.StateToCommandsMap[stateExecutionIdKey] = append(s.StateToCommandsMap[stateExecutionIdKey][:i],
+				s.StateToCommandsMap[stateExecutionIdKey][i+1:]...)
+			if len(s.StateToCommandsMap[stateExecutionIdKey]) == 0 {
+				delete(s.StateToCommandsMap, stateExecutionIdKey)
+			}
+
+			return &stateExecutionIdKey, consumedDedupIds
+		}
+	}
+
+	return nil, []uuid.UUID{}
+}
+
+// ConsumeFor return a bool indicating if the stateExecutionId can complete the local queue commands, and an array of all the consumed dedupIds.
+//
+// E.g., given UnconsumedMessageQueueCountMap as:
+//
+// (q1, 2), (q2, 2), (q3, 1)
+//
+// and StateToCommandsMap[stateExecutionId] as:
+//
+// (q1, 1), (q2, 2)
+//
+// and isAllOfCompletion as true. Then after ConsumeFor, the UnconsumedMessageQueueCountMap becomes:
+//
+// (q1, 1), (q3, 1)
+//
+// and returns:
+//
+// (true, [q1_dedup_id_1, q2_dedup_id_1, q2_dedup_id_2])
+func (s *StateExecutionWaitingQueuesJson) ConsumeFor(stateExecutionId StateExecutionId, isAllOfCompletion bool) (bool, []uuid.UUID) {
```

Comment:
Nit:
* Maybe `ConsumeFor` --> `CheckCanCompleteLocalQueueWaiting` (it's a bit long, but much clear by just reading the name)
* `isAllOfCompletion`, maybe just use xdbapi.CommandWaitingType enum instead. We will have others in the future.